app/notification_email.go

```diff
@@ -284,6 +288,23 @@ func getFormattedPostTime(user *model.User, post *model.Post, useMilitaryTime bo
 	}
 }
 
+func (a *App) generateHyperlinkForChannels(postMessage, teamName, teamURL string) string {
+	r := regexp.MustCompile(`\~[-a-zA-Z0-9]+\b`)
+	matches := r.FindAllString(postMessage, -1)
+	var channelURL, channelHyperLink string
+	for _, channelName := range matches {
+		ch, err := a.GetChannelByNameForTeamName(channelName[1:], teamName, true)
```

Comment:
This is going to leak some data about whether or not a given channel name exists. For example, if I could post a link to `~firing-joe`, and if it is turned into a link, then I know that there's a channel about firing Joe even if it's private and I can't see it myself.

As far as I know, our stance on that is that this is confidential data, so we might need some logic here to not link private channels, although that would really complicate this since then the email text would be based off of who is receiving the email which would heavily change this code. I'm going to bring this up with our security team to get a firm decision on this since while this has been our stance in the past, there are other ways to get this information.

Reply:
We're going to have to change how this works since we do consider channel names confidential, and we can't customize the email for different users with how email notifications are set up.

@esethna How would you feel if we just turn anything that looks like a channel link into a link? It's obviously not ideal since a user might make a broken link by accident or if they make a typo, but it would at least fix the issue that those links aren't clickable. Alternatively, we could only link public channels in emails which fixes the problem of broken links, but it prevents private channel links from working.

Reply:
Sorry, missed this notification.

@hmhealey @waseem18 given the above security concern I think **we should only be linking Public channels**. Reason being that if we link everything that looks like a channel name we might link things such as `~3 hours from now` or `around ~thursday evening`

Reply:
Sure @esethna